### PR TITLE
Add ST_Length(SphericalGeography)

### PR DESCRIPTION
**Pull request by seanbarker**
Adds `ST_Length(SphericalGeography)` which computes the length of a `LINESTRING` on the surface of a spherical model of Earth's surface, i.e. the sum of great-circle distances between adjacent points along the linestring.

### CODE REVIEW

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 1]
```diff
@@ -201,6 +201,59 @@ public void testArea()
...
     private void assertArea(String wkt, double expectedArea)
     {
-        assertFunction(format("ABS(ROUND((ST_Area(to_spherical_geography(ST_GeometryFromText('%s'))) / %f - 1 ) * %d, 0))", wkt, expectedArea, 10000), DOUBLE, 0.0);
+        assertFunctionImprecise(format("ST_Area(to_spherical_geography(ST_GeometryFromText('%s')))", wkt), expectedArea, 1e-4);
     }
+
+    @Test
+    public void testLength()
+    {
+        double x1 = -71.05, y1 = 42.36,   // Boston
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: declare each variable separately

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 2]
```diff
@@ -201,6 +201,59 @@ public void testArea()
 
     private void assertArea(String wkt, double expectedArea)
     {
-        assertFunction(format("ABS(ROUND((ST_Area(to_spherical_geography(ST_GeometryFromText('%s'))) / %f - 1 ) * %d, 0))", wkt, expectedArea, 10000), DOUBLE, 0.0);
+        assertFunctionImprecise(format("ST_Area(to_spherical_geography(ST_GeometryFromText('%s')))", wkt), expectedArea, 1e-4);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: Why did you make this change? Seems unrelated.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 3]
```diff
@@ -201,6 +201,59 @@ public void testArea()
...
+
+        // Linestring with one point has length 0
+        assertLength(format("LINESTRING (%.2f %.2f)", x1, y1),0.0);
+
+        // Linestring with only one distinct point has length 0
+        assertLength(format("LINESTRING (%.2f %.2f, %.2f %.2f, %.2f %.2f)", x1, y1, x1, y1, x1, y1), 0.0);
+
+        // ST_Length is equivalent to sums of ST_DISTANCE between points in the LineString
+        assertDistance(format("POINT (%.2f %.2f)", x1, y1), format("POINT (%.2f %.2f)", x2, y2), distance1);
+        assertDistance(format("POINT (%.2f %.2f)", x2, y2), format("POINT (%.2f %.2f)", x3, y3), distance2);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: these tests don't belong here; this test should only be testing ST_Length; these is a separate test for ST_Distance

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 4]
```diff
@@ -201,6 +201,59 @@ public void testArea()
...
+        assertLength(format("LINESTRING (%.2f %.2f, %.2f %.2f, %.2f %.2f)", x1, y1, x2, y2, x3, y3), length);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: I find this difficult to read. Hardcoding coordinates and expected length will be clearer.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 5]
```diff
@@ -201,6 +201,59 @@ public void testArea()
...
+               x3 = -122.41, y3 = 37.77;  // San Francisco
+
+        double distance1 = 1365691.096620299, distance2 = 2985175.539584118;
+        double length = distance1 + distance2;
+
+        // Empty linestring returns null
+        assertLength("LINESTRING EMPTY", null);
+
+        // Linestring with one point has length 0
+        assertLength(format("LINESTRING (%.2f %.2f)", x1, y1),0.0);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: No need to use any special values here. LINESTRING (0 0) would work just fine.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 6]
```diff
@@ -201,6 +201,59 @@ public void testArea()
...
+
+        // Linestring with only one distinct point has length 0
+        assertLength(format("LINESTRING (%.2f %.2f, %.2f %.2f, %.2f %.2f)", x1, y1, x1, y1, x1, y1), 0.0);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: LINESTRING (0 0, 0 0, 0 0) would be clearer

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 7]
```diff
@@ -201,6 +201,59 @@ public void testArea()
...
+        assertDistance(format("POINT (%.2f %.2f)", x1, y1), format("POINT (%.2f %.2f)", x2, y2), distance1);
+        assertDistance(format("POINT (%.2f %.2f)", x2, y2), format("POINT (%.2f %.2f)", x3, y3), distance2);
+        assertLength(format("LINESTRING (%.2f %.2f, %.2f %.2f, %.2f %.2f)", x1, y1, x2, y2, x3, y3), length);
+
+        // Linestring has same length as its reverse
+        assertLength(format("LINESTRING (%.2f %.2f, %.2f %.2f, %.2f %.2f)", x3, y3, x2, y2, x1, y1), length);
+
+        // Distance between north and south pole should be roughly half of the circumference of the Earth
+        double earthCircumference = 4.0075e7; // in meters
+        assertLength(format("LINESTRING (%.2f %.2f, %.2f %.2f)", 0.0, 90.0, 0.0, -90.0), earthCircumference / 2, 1e-2);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: better hard-code the expected distance and hardcoded the WKT

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 8]
```diff
@@ -201,6 +201,59 @@ public void testArea()
 
     private void assertArea(String wkt, double expectedArea)
     {
-        assertFunction(format("ABS(ROUND((ST_Area(to_spherical_geography(ST_GeometryFromText('%s'))) / %f - 1 ) * %d, 0))", wkt, expectedArea, 10000), DOUBLE, 0.0);
+        assertFunctionImprecise(format("ST_Area(to_spherical_geography(ST_GeometryFromText('%s')))", wkt), expectedArea, 1e-4);
     }
+
+    @Test
+    public void testLength()
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: Could you add a test for a circular linestring?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 9]
```diff
@@ -203,4 +203,63 @@ private void assertArea(String wkt, double expectedArea)
     {
         assertFunction(format("ABS(ROUND((ST_Area(to_spherical_geography(ST_GeometryFromText('%s'))) / %f - 1 ) * %d, 0))", wkt, expectedArea, 10000), DOUBLE, 0.0);
     }
+
+    @Test
+    public void testLength()
+    {
+        double distance1 = 1365691.096620299;
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: distance1 and distance2 variables are used only once to compute length; it is clearer to use assign length to a pre-computed constant value.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 10]
```diff
@@ -203,4 +203,63 @@ private void assertArea(String wkt, double expectedArea)
...
+    {
+        assertFunctionImprecise(format("ST_Length(to_spherical_geography(ST_GeometryFromText('%s')))", lineString), expectedLength, tolerance);
+    }
+
+    private void assertLength(String lineString, Double expectedLength)
+    {
+        assertLength(lineString, expectedLength, 1e-4);
+    }
+
+    private void assertFunctionImprecise(String function, Double expected, double tolerance)
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: this function is used only once, let's inline it

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 11]
```diff
@@ -203,4 +203,63 @@ private void assertArea(String wkt, double expectedArea)
...
+        assertLength("LINESTRING (0 0, 0 0, 0 0)", 0.0);
+
+        // ST_Length is equivalent to sums of ST_DISTANCE between points in the LineString
+        assertLength("LINESTRING (-71.05 42.36, -87.62 41.87, -122.41 37.77)", length);
+
+        // Linestring has same length as its reverse
+        assertLength("LINESTRING (-122.41 37.77, -87.62 41.87, -71.05 42.36)", length);
+
+        // Path north pole -> south pole -> north pole should be roughly the circumference of the Earth
+        assertLength("LINESTRING (0.0 90.0, 0.0 -90.0, 0.0 90.0)", 4.0075e7, 1e-2);
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: Would it be possible to use the default tolerance here?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java (hunk 8 above)
Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: @seanbarker Sean, I don't see a test for a circular linestring. Could you give me a pointer?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java (hunk 11 above)
Comment (seanbarker) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: The test fails with the default tolerance, but the difference in the output is only about 44 km. I can either change the estimate of earth's circumference and use the default tolerance or use the lower tolerance.

Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: Let's change the estimate.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 12]
```diff
@@ -203,4 +203,53 @@ private void assertArea(String wkt, double expectedArea)
...
+
+    private void assertLength(String lineString, Double expectedLength)
+    {
+        String function = format("ST_Length(to_spherical_geography(ST_GeometryFromText('%s')))", lineString);
+
+        if (expectedLength == null || expectedLength == 0.0) {
+            assertFunction(function, DOUBLE, expectedLength);
+        }
+        else {
+            // assert the function returns the expected value, within the given error tolerance
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: this comment is redundant, it restates the code in the next line; please, remove

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java [hunk 13]
```diff
@@ -203,4 +203,53 @@ private void assertArea(String wkt, double expectedArea)
     {
         assertFunction(format("ABS(ROUND((ST_Area(to_spherical_geography(ST_GeometryFromText('%s'))) / %f - 1 ) * %d, 0))", wkt, expectedArea, 10000), DOUBLE, 0.0);
     }
+
+    @Test
+    public void testLength()
+    {
+        double length = 4350866.6362;
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestSphericalGeoFunctions.java]: nit: move this variable closer to its first usage

**Comment by mbasmanova** [comment]
@seanbarker Sean, thank you for the contribution.
